harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,412 @@
+import abc
+import logging
+import queue
+import threading
+import time
+from datetime import datetime, timezone
+from typing import Any, Dict, List
+
+import psutil
+
+from determined import core
+from determined.common import api, constants
+from determined.common.api import bindings
+
+try:
+    import pynvml
+except ImportError:
+    pynvml = None
+
+logger = logging.getLogger("determined.core")
+
+
+class ProfilerContext:
+    """
+    ``ProfilerContext`` gives access to the system profiling feature within Determined.
+
+    It is responsible for collecting system metrics at specified time intervals and reporting
+    them to the master.
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        distributed: core.DistributedContext,
+    ) -> None:
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+        self._distributed = distributed
```

Comment:
I don't want the a `ProfilerContext` to store a `DistributedContext`. That sounds more complicated than it needs. Is there an alternative design that makes this class simpler?

Reply:
I'm not very educated when it comes to Contexts. So far, the prejudice I have is that they're a place to store global variables so we don't have to do the hard work of keeping our abstractions neat.

I think that's a little bit how `DistributedContext` is used here, too. If only rank is important, than I think a `ProfilerContext` should take a `rank`. If that is difficult or ugly, then maybe that's a sign that there's a better abstraction out there.

Actually... I _don't_ fully think that a `ProfilerContext` should take a `rank`. I instead would most prefer it only gets created on the chief. That feels cleaner.

Maybe a chat here would be best, so you can explain to me why that's a bad idea?

Reply:
I'm still not totally comfortable with the profiler itself having to know greater context. Profiler gonna profile, let the object that instantiates it be responsible for who should profile and when.

If we decide this is the best design, I'd still like to limit the amount of information available to the profiler -- for it to take a "jk don't profile" flag or something instead of making the profiler itself be responsible for orchestration.

Reply:
We've decided that this is in fact the most consistent way of passing information that the profiler needs. We've also decided that the profiler needs this information because we don't want to give Controllers total responsibility for actually orchestrating profiling. Just some responsibility.